cert.go

```diff
@@ -343,6 +345,8 @@ func (m *mkcert) newCA() {
 		BasicConstraintsValid: true,
 		IsCA:                  true,
 		MaxPathLenZero:        true,
+
+		PermittedDNSDomains: m.dnsWhitelist,
```

Comment:
See https://github.com/caddyserver/caddy/issues/5759#issuecomment-1690700681

> People using name constraints should know what they exactly mean, as some cases are not obvious. For example, **adding just `permittedDNSDomains` as above does not exclude creating domains with IP addresses or any other type of SAN**. Name constraints are defined in [RFC5280#4.2.1.10](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10)

Would it be possible to prevent the cert to be valid for those other things when the whiltelist option is given for a _domain_?

Reply:
Feel free to suggest the set of changes required to improve security, I don't have resources to spend on this PR myself.

Reply:
@ivanbakel Oh hi Isaac! I just now realised who filed the PR.

No problem from my side, I just encountered this question when doing research for https://github.com/nh2/internal-contstrained-pki. I don't know the solution yet.

Reply:
After reading the spec it seems like name constraints can be specified on DNS names, directory names, e-mail addresses, IP ranges and URIs. They can be permissions and exclusions. Exclusions have priority.

So to disallow issuing certificates with IP range names, we should add _all_ IP addresses (I guess 0.0.0.0 with subnet mask 0.0.0.0, and also IPv6 [::]/0) to the `ExcludedIPRanges` argument: https://pkg.go.dev/crypto/x509#CreateCertificate

If I read the RFC correctly URI name constraints cannot be used to later issue certificates with IP address or wildcard host URIs, so one would just duplicate the DNS restriction as URI restriction (I guess for https, but one could make it configurable if other protocols are needed).

If S/MIME is desired one can add the DNS restriction as host for mailbox restrictions.

However I am not sure how to disallow e-mail addresses or directory names in general.

If anyone knows more, I'd be interested

Reply:
@delreluca  I hope my comment addresses some of these questions. The openssl syntax is quite easy to follow [here](https://docs.openssl.org/master/man5/x509v3_config/#name-constraints), with [this](https://docs.openssl.org/master/man5/x509v3_config/#subject-alternative-name), so it will be a good starting point if you want to check the output certificate and work backwards from there to adapt it to `go`